game-app/game-headed/src/main/java/games/strategy/engine/framework/startup/ui/ServerSetupPanel.java

```diff
@@ -334,44 +336,84 @@ private void internalPlayerListChanged() {
     final Map<String, Boolean> playersEnabled = model.getPlayersEnabledListing();
     final Map<String, String> reloadSelections =
         GamePlayer.currentPlayers(gameSelectorModel.getGameData());
+    final List<PlayerTypes.Type> visiblePlayerTypes =
+        CollectionUtils.getMatches(HeadedPlayerTypes.getPlayerTypes(), PlayerTypes.Type::isVisible);
     for (final Map.Entry<String, Collection<String>> entry :
         model.getPlayerNamesAndAlliancesInTurnOrder().entrySet()) {
-      final PlayerRow newPlayerRow =
-          new PlayerRow(
-              entry.getKey(),
-              reloadSelections,
-              entry.getValue(),
-              new PlayerTypes(HeadedPlayerTypes.getPlayerTypes()));
-      playerRows.add(newPlayerRow);
-      newPlayerRow.update(players, playersEnabled);
+      final String playerName = entry.getKey();
+      GamePlayer player = null;
+      if (gameSelectorModel.getGameData() != null) {
+        player = gameSelectorModel.getGameData().getPlayerList().getPlayerId(playerName);
+      }
+      PlayerTypes.Type playerType =
+          determinePlayerType(playerName, player, reloadSelections, visiblePlayerTypes);
+      PlayerRow row = new PlayerRow(playerName, playerType, visiblePlayerTypes, entry.getValue());
+      row.update(players, playersEnabled);
+      playerRows.add(row);
     }
     layoutPlayers();
     internalPlayersTakenChanged();
   }
 
+  private PlayerTypes.Type determinePlayerType(
+      String playerName,
+      @Nullable GamePlayer player,
+      Map<String, String> reloadSelections,
+      List<PlayerTypes.Type> visiblePlayerTypes) {
+    final String previousSelection = reloadSelections.get(playerName);
+    if (previousSelection != null && previousSelection.equalsIgnoreCase("Client")) {
+      return HeadedPlayerTypes.HUMAN_PLAYER;
+    } else if (previousSelection != null && !previousSelection.equals("no_one")) {
```

Comment:
Do you have a clue where these string constants are coming from? If yes, could you add a comment for future reference?

Reply:
"no_one" comes from the default value of `whoAmI` in GamePlayer, but I'm not sure where "Client" comes from. I'll add at least a comment about the former.